File: src/main/java/com/tms/csp/transforms/OrWithNegAndArg.java
package com.tms.csp.transforms;

import com.google.common.collect.ImmutableList;
import com.tms.csp.ast.Exp;
import com.tms.csp.ast.ExpSetOld;

import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;

//or(a b c !and(x y z)
// to: or(a b c or(!x !y !z))
// to: or(a b c !x !y !z)


/**
 *  or(a b c !and(x y z)
 *  to: or(a b c or(!x !y !z))
 *  to: or(a b c !x !y !z)
 */
public class OrWithNegAndArg extends BaseTransformer {

    @Override
    protected boolean executeLocal(Exp in) {
        return in.isOr() && hasNegAndArg(in);
    }

    @Override
    protected Exp local(Exp in) {
        checkArgument(in.isOr() && hasNegAndArg(in));
        int i = indexOfNegAndArg(in);

        Exp negAndArg = in.getArg(i);
        assert negAndArg.isNegAnd();


        ImmutableList<Exp> argsCopy = in.copyOfArgs();

        throw new UnsupportedOperationException();
    }

//    static Exp convertNegAndToOr(Exp negAnd) {
//        assert negAnd.isNegAnd();
//        Exp arg = negAnd.getArg();
//        List<Exp> args = arg.getArgs();
//        List<Exp> flippedArgs = ExpSetOld.flipAll(args);
//        return negAnd.mkOr(flippedArgs);
//    }

    boolean hasNegAndArg(Exp in) {
        return indexOfNegAndArg(in) != -1;
    }

    int indexOfNegAndArg(Exp in) {
        List<Exp> args = in.args();
        for (int i = 0; i < args.size(); i++) {
            Exp e = args.get(i);
            if (e.isNegAnd()) {
                return i;
            }
        }
        return -1;
    }


}

File: src/main/java/com/tms/csp/transforms/OrWithNegAndArg.kt
package com.tms.csp.transforms

import com.google.common.collect.ImmutableList
import com.tms.csp.ast.Exp
import com.tms.csp.ast.ExpSetOld

import com.google.common.base.Preconditions.checkArgument

//or(a b c !and(x y z)
// to: or(a b c or(!x !y !z))
// to: or(a b c !x !y !z)


/**
 * or(a b c !and(x y z)
 * to: or(a b c or(!x !y !z))
 * to: or(a b c !x !y !z)
 */
class OrWithNegAndArg : BaseTransformer() {

    override fun executeLocal(inExp: Exp): Boolean {
        return inExp.isOr && hasNegAndArg(inExp)
    }

    override fun local(inExp: Exp): Exp? {
        checkArgument(inExp.isOr && hasNegAndArg(inExp))
        val i = indexOfNegAndArg(inExp)

        val negAndArg = inExp.getArg(i)
        assert(negAndArg.isNegAnd)


        val argsCopy = inExp.copyArgs()

        throw UnsupportedOperationException()
    }

    //    static Exp convertNegAndToOr(Exp negAnd) {
    //        assert negAnd.isNegAnd();
    //        Exp arg = negAnd.getArg();
    //        List<Exp> args = arg.getArgs();
    //        List<Exp> flippedArgs = ExpSetOld.flipAll(args);
    //        return negAnd.mkOr(flippedArgs);
    //    }

    internal fun hasNegAndArg(inExp: Exp): Boolean {
        return indexOfNegAndArg(inExp) != -1
    }

    internal fun indexOfNegAndArg(inExp: Exp): Int {
        val args = inExp.args
        for ((i,e) in args.withIndex()) {
            if (e.isNegAnd) {
                return i
            }
        }
        return -1
    }


}
